feat: Add new GHA to create urgent hotfix images

szpraat opened the pull request:
# 🔥 Create Urgent Hotfix Images GitHub Actions Workflow

## 🎯 **Objective**

This PR introduces a new GitHub Actions workflow to automate the creation and deployment of urgent hotfix Docker images for Camunda components. The workflow streamlines the emergency hotfix process by eliminating manual Docker builds and ensuring consistent image tagging and registry management.

## 🚀 **Key Features**

### **Automated Hotfix Image Creation**
- **Manual Trigger**: Workflow dispatch with intuitive input parameters
- **Multi-Component Support**: Build multiple components in parallel (`camunda`, `operate`, `tasklist`, `optimize`,`zeebe`)
- **Flexible Source**: Build from either branch or specific commit SHA
- **Registry Integration**: Automatic push to `registry.camunda.cloud/hotfixes`
- **Multi-Platform Images**: Automatically builds for both `linux/amd64` and `linux/arm64` architectures

### **Robust Input Validation**
- **Component Validation**: Only allows supported Camunda components
- **Version Format Validation**: Supports semver and snapshot versions
- **Support Ticket Integration**: Required ticket ID for proper tracking
- **Git Reference Validation**: Ensures branch/commit exists before build

### **Production-Ready Architecture**
- **Multi-Stage Validation**: Input → Build → Verification → Push
- **Matrix Strategy**: Parallel builds for multiple components
- **Proper Error Handling**: Fail-fast with clear error messages
- **Comprehensive Logging**: Detailed build progress and artifact verification

## 🏗️ **Technical Implementation**

### **Repository-Aligned Build Process**
- **Maven Integration**: Uses proper distribution modules (`-pl dist`, `-pl operate-distro`, etc.)
- **Dockerfile Detection**: Automatic selection of component-specific or main Dockerfile
- **Build Optimization**: Multi-threaded builds with `-T1C` and skip flags for speed
- **Artifact Verification**: Confirms distribution tarballs exist before Docker build

### **Security & Authentication**
- **HashiCorp Vault Integration**: Retrieves Harbor credentials securely
- **Minimal Permissions**: Uses least-privilege access patterns
- **Secret Management**: No hardcoded credentials, follows Camunda security guidelines

### **Docker Build Standards**
- **Multi-Platform Support**: Linux/amd64 and linux/arm64 platform
- **OCI Compliance**: Proper labels and metadata
- **Caching**: GitHub Actions cache for faster builds
- **Naming Convention**: Follows Camunda hotfix standards

## 📋 **Workflow Usage**

### **Input Format**
```yaml
# Components with versions
components_with_versions: "camunda:8.6.1,operate:8.6.1"

# Support ticket (required)
support-ticket-id: "SUPPORT-12345"

# Source (either branch OR commit)
branch: "hotfix/security-fix"

# OR
commit: "abc123def"
```
### **Output Images**
```yaml
# Primary tag
registry.camunda.cloud/hotfixes/camunda:8.6.1-support-12345
# SHA tag (for exact reproducibility)
registry.camunda.cloud/hotfixes/camunda:8.6.1-support-12345-abc12345
```

### Testing

#### Input with branch: 
 `
BRANCH="test-hotfix-workflow-8.7.15"
COMMIT=""
TICKET_ID="TEST-WORKFLOW-12345"
COMPONENTS="camunda:8.7.15-TEST"
`
<img width="1201" height="623" alt="image" src="https://github.com/user-attachments/assets/edc5fe6e-79ea-4670-a41f-9f585df17a3d" />

Created img in Harbor:
<img width="1229" height="412" alt="image" src="https://github.com/user-attachments/assets/87b25c97-461b-471c-8ec4-0e19807f47fc" />
#### Input with commit sha:
'   BRANCH=""
   COMMIT="c319e3a3fa5"
   TICKET_ID="TEST-WORKFLOW-12345"
   COMPONENTS="camunda:8.7.15-TEST"'
 
<img width="1173" height="636" alt="image" src="https://github.com/user-attachments/assets/c6ada96e-df69-456d-9179-a8ae35df70a3" />

### For optimize:
  ` BRANCH=""
   COMMIT="2ed56a97c07"
   TICKET_ID="TEST-WORKFLOW-12345"
   COMPONENTS="optimize:8.8.0-TEST"`
   
   
<img width="1381" height="396" alt="image" src="https://github.com/user-attachments/assets/2520254b-94ab-4e45-81fc-643081edd311" />

### For camunda,tasklist and operate all at once:
   `BRANCH=""
COMMIT="2ed56a97c07"
TICKET_ID="TEST-WORKFLOW-12345"
COMPONENTS="camunda:8.8.0-TEST,tasklist:8.8.0-TEST,operate:8.8.0-TEST"`

<img width="1022" height="1096" alt="image" src="https://github.com/user-attachments/assets/67d73504-9975-4f89-940f-4ef83cd798bf" />
<img width="681" height="371" alt="image" src="https://github.com/user-attachments/assets/438abf7b-de62-4319-9312-23cf06f051d1" />
<img width="678" height="405" alt="image" src="https://github.com/user-attachments/assets/39081e1d-e952-4e3b-8168-66c97b73a18e" />
<img width="689" height="375" alt="image" src="https://github.com/user-attachments/assets/9f4445ee-eac9-45a6-a3f0-cb7722c08ca3" />

### Multiplatform images:
<img width="912" height="338" alt="image" src="https://github.com/user-attachments/assets/5dddbbf6-d594-4e78-969b-53547d977cb2" />

### Related issue
https://github.com/camunda/camunda/issues/31357

szpraat committed:
feat: Add GitHub Action for urgent hotfix image creation

- Implement manual trigger workflow for creating urgent hotfix Docker images
- Add matrix strategy for parallel builds of camunda, zeebe, operate, tasklist, optimize
- Include comprehensive input validation for commit SHAs and ticket IDs
- Add multi-platform Docker builds (linux/amd64, linux/arm64)
- Integrate Maven settings with Camunda Nexus mirror configuration
- Add SBOM generation and provenance attestation for security compliance
- Include proper image labeling for hotfix tracking and identification

maxdanilov reviewed (changes requested):
Looks good overall, but I think the code can be simplified in many places: we should aim to keep the first iteration as basic as possible.

szpraat commented on .github/workflows/create-urgent-hotfix-img.yml:
```diff
@@ -0,0 +1,541 @@
+      - name: Setup Maven cache
+        uses: ./.github/actions/setup-maven-cache
+        with:
+          maven-cache-key-modifier: hotfix
```
In cases when for example multiple modules are built like tasklist, operate and optimize at the same time. In that case it can be useful. I can remove it, as its an edge case what I wanted to cover.

szpraat commented on .github/workflows/create-urgent-hotfix-img.yml:
```diff
@@ -0,0 +1,541 @@
+
+          # Validate ticket ID format (should be alphanumeric with hyphens/underscores)
+          if [[ ! "$TICKET_ID" =~ ^[a-zA-Z0-9_-]+$ ]]; then
+            echo "❌ Error: Support ticket ID contains invalid characters. Use only alphanumeric, hyphens, and underscores"
```
Added in commit 794e4291944e91879da18cc3f0857935bac13652

szpraat commented on .github/workflows/create-urgent-hotfix-img.yml:
```diff
@@ -0,0 +1,541 @@
+          echo "Matrix JSON: $MATRIX_JSON"
+          echo "matrix=$MATRIX_JSON" >> "$GITHUB_OUTPUT"
+      - name: Checkout repository to determine ref
+        uses: actions/checkout@v4
```
Updated in commit 794e4291944e91879da18cc3f0857935bac13652

szpraat commented on .github/workflows/create-urgent-hotfix-img.yml:
```diff
@@ -0,0 +1,541 @@
+          case "${{ matrix.component }}" in
+            "camunda")
+              echo "📦 Building camunda (includes zeebe, operate, tasklist) distribution tarball..."
+              ./mvnw clean package -T1C -DskipTests -DskipChecks -am -pl dist
```
Simplified in commit 794e4291944e91879da18cc3f0857935bac13652

szpraat commented on .github/workflows/create-urgent-hotfix-img.yml:
```diff
@@ -0,0 +1,541 @@
+            "camunda")
+              echo "📦 Building camunda (includes zeebe, operate, tasklist) distribution tarball..."
+              ./mvnw clean package -T1C -DskipTests -DskipChecks -am -pl dist
+              echo "✅ Built: dist/target/camunda-zeebe-*.tar.gz"
```
It was just to log the process of the build. Removed in commit 794e4291944e91879da18cc3f0857935bac13652

maxdanilov commented on .github/workflows/create-urgent-hotfix-img.yml:
```diff
@@ -0,0 +1,541 @@
+        required: false
+        type: string
+      commit:
+        description: 'The specific commit SHA to create hotfix images from (accepts both short and full SHA - takes precedence over branch)'
```
In general, I'm fine for this support both branch name and commits, but not really happy to see so much workflow code is required to support it with all these reference lookup operations.

Can we drastically simplify the supporting code? It can be the user needs to provide more data to the workflow, that would be fine.

In my ideal world, if we provide branch name, the workflow just does `git checkout <commit>` or `git checkout -b <branch>` and then does build operations on that state. Can we achieve it somehow?

maxdanilov replied:
or even this: when triggering a manual start of the hotfix workflow, the branch can be selected - that GitHub Actions takes care of the check-out, wdyt?

<img width="333" height="365" alt="Screenshot 2025-10-21 at 17 57 28" src="https://github.com/user-attachments/assets/a48616d4-e985-4c5f-a0b8-edf682a01e1c" />

maxdanilov commented on .github/workflows/create-urgent-hotfix-img.yml:
```diff
@@ -0,0 +1,541 @@
+        run: |
+          git fetch --all --tags --force
+          git fetch origin '+refs/heads/*:refs/remotes/origin/*' --force
+      - name: Determine reference and SHA
```
It's also fine to challenge the requirements (and align with the stakeholders) if they (requirements) make the implementation significantly harder and are not must haves. Let's continue this discussion in https://github.com/camunda/camunda/pull/39491#discussion_r2448839412.

maxdanilov commented on .github/workflows/create-urgent-hotfix-img.yml:
```diff
@@ -0,0 +1,301 @@
+        id: parse
+        run: |
+          BRANCH="${{ github.event.inputs.branch }}"
+          COMMIT="${{ github.event.inputs.commit || 'f7ff978967f6ccd4143b9b0450a5f0aa036dd4ba' }}"
```
(reminder) This fallback value is to be removed before merging.

maxdanilov commented on .github/workflows/create-urgent-hotfix-img.yml:
```diff
@@ -0,0 +1,301 @@
+            if [[ "$ITEM" =~ ^([^:]+):(.+)$ ]]; then
+              COMP="${BASH_REMATCH[1]}"
+              case "$COMP" in
+                zeebe|camunda|optimize|operate|tasklist) VALID_COMPONENTS+=("$ITEM") ;;
```
```suggestion
                camunda|optimize|operate|tasklist|zeebe)
```
(alphabetical order for readability)

maxdanilov commented on .github/workflows/create-urgent-hotfix-img.yml:
```diff
@@ -0,0 +1,301 @@
+          echo "🔨 Building Maven artifacts for ${{ matrix.component }}..."
+
+          case "${{ matrix.component }}" in
+            "zeebe"|"camunda"|"operate"|"tasklist")
```
alphabetical recommended here as well for readability:

```suggestion
            "camunda"|"operate"|"tasklist"|"zeebe")
```

maxdanilov commented on .github/workflows/create-urgent-hotfix-img.yml:
```diff
@@ -0,0 +1,301 @@
+# - Platforms are not configurable, they are by default set to linux/amd64 and linux/arm64
+#
+# The workflow validates inputs, builds the requested images, and pushes them to the Camunda hotfix registry following
+# the naming convention: registry.camunda.cloud/hotfixes/<image>:<application-version>-<support-issue-id>
```
to be updated after https://github.com/camunda/camunda/pull/39491/files#r2452690290

maxdanilov commented on .github/workflows/create-urgent-hotfix-img.yml:
```diff
@@ -0,0 +1,301 @@
+          file: ${{ matrix.component == 'zeebe' && 'Dockerfile' || format('{0}.Dockerfile', matrix.component) }}
+          platforms: ${{ env.PLATFORMS }}
+          push: true
+          sbom: true
```
:x: 

why SBOM is being pushed here?I don't see that being done in docker places using docker/build-push-action.

szpraat commented on .github/workflows/create-urgent-hotfix-img.yml:
```diff
@@ -0,0 +1,301 @@
+            DATE=$BUILD_DATE
+            REVISION=${{ steps.resolve.outputs.ref }}
+            DISTRO=production
+            ARTIFACT_PATH=./optimize-distro/target"
```
Wanted to keep everything in one place, even if it has default value. To see every related default values in one place.

szpraat commented on .github/workflows/create-urgent-hotfix-img.yml:
```diff
@@ -0,0 +1,301 @@
+            DISTBALL=dist/target/camunda-zeebe-*.tar.gz"
+          fi
+
+          # Use printf to preserve multiline format and escape newlines for output
```
Yes, accidentally left there, will be removed

szpraat commented on .github/workflows/create-urgent-hotfix-img.yml:
```diff
@@ -0,0 +1,301 @@
+            "optimize")
+              echo "📦 Building optimize distribution tarball..."
+              if [[ ! -f "optimize.Dockerfile" ]]; then
+                echo "❌ ERROR: Optimize not available in this version (only supported from 8.8+)"
```
Change commited in 68f072f2b22df573ac732fa34ced6bbf32392d29

szpraat committed:
feat: Add Docker authentication and production readiness

- Add authentication for Harbor registry (registry.camunda.cloud)
- Add authentication for Minimus registry (reg.mini.dev) for internal base images
- Add conditional Maven profile (-PrunAssembly) for optimize component compatibility
- Remove test default values and artifacts for production deployment
- Validate workflow across multiple stable branches (8.6, 8.7, 8.8, optimize variants)
- Ensure production-ready configuration with proper error handling

maxdanilov commented on .github/workflows/create-urgent-hotfix-img.yml:
```diff
@@ -0,0 +1,296 @@
+                echo "❌ ERROR: Optimize not available in this version, make sure your source branch includes it."
+                exit 1
+              fi
+              ./mvnw clean package -T1C -DskipTests -DskipChecks -am -pl optimize-distro
```
nice, thank you.

maxdanilov reviewed (changes requested):
Overall LGTM, but I just noticed there's potential to reuse the existing setup action to make the workflow lighter. Sorry I didn't catch it earlier, only occured to me now.

szpraat committed:
feat: Add .github/actions/setup-build and test it

szpraat commented on .github/workflows/create-urgent-hotfix-img.yml:
```diff
@@ -0,0 +1,313 @@
+            secret/data/github.com/organizations/camunda NEXUS_USR | ci-account-username;
+            secret/data/products/camunda/ci/github-actions REGISTRY_MINIMUS_PSW | minimus-token;
+      
+      - name: Log into Harbor
```
[main](https://github.com/camunda/camunda/actions/runs/18939116823/attempts/1#summary-54073976573)
<img width="1395" height="1077" alt="image" src="https://github.com/user-attachments/assets/7c6d6985-dc98-430d-a958-b0eb0b78e345" />

[stable/8.8](https://github.com/camunda/camunda/actions/runs/18938608272/attempts/1#summary-54072290835)
<img width="1119" height="918" alt="image" src="https://github.com/user-attachments/assets/1c40362e-d267-4ce7-90ba-f94f9a7f3b13" />


[stable/optimize-8.6](https://github.com/camunda/camunda/actions/runs/18939207110/attempts/1#summary-54074198906)
<img width="1255" height="1073" alt="image" src="https://github.com/user-attachments/assets/e126349c-bbad-4b4f-8f89-cf18850d8b47" />

maxdanilov reviewed (approved):
LGTM, thank you!
